### PR TITLE
#161348809 Add an article read time

### PR DESCRIPTION
**Pull request by dnuwa**
#### What does this PR do?
Adds a read time to an article

#### Description of Task to be completed?
Add read time to each article automatically

#### How should this be manually tested?
- Retrieve all articles
GET `/api/articles`

- Retrieve a single article
GET `/api/articles/11`

Example response:
```
{
    "id": 11,
    "tag_list": [
        "science",
        "technology"
    ],
    "slug": "how-to-train-your-dragon1542180385794015",
    "title": "How to train your dragon",
    "description": "Ever wonder how?",
    "body": "The story takes place in a mythical Viking world where a young Viking teenager named Hiccup aspires to follow his tribe's tradition of becoming a dragon slayer. The story takes place in a mythical Viking world where a young Viking teenager named Hiccup aspires to follow his tribe's tradition of becoming a dragon slayer. The story takes place in a mythical Viking world where a young Viking teenager named Hiccup aspires to follow his tribe's tradition of becoming a dragon slayer. The story takes place in a mythical Viking world where a young Viking teenager named Hiccup aspires to follow his tribe's tradition of becoming a dragon slayer. The story takes place in a mythical Viking world where a young Viking teenager named Hiccup aspires to follow his tribe's tradition of becoming a dragon slayer.",
    "created_at": "2018-11-14T07:26:25.855564Z",
    "updated_at": "2018-11-14T07:26:25.855583Z",
    "image_url": "http://iviidev.info/downloads/image.jpg",
    "audio_url": null,
    "read_time": 1,
    "author": {
        "id": 1,
        "username": "daniel",
        "email": "danielnuwa84@gmail.com"
    },
    "likes": 0,
    "dislikes": 0,
    "Favorites_count": 0,
    "rating": 0
}
```

#### Any background context you want to provide?
The read time is calculated automatically.
Each article is returned with its read time.

#### What are the relevant pivotal tracker stories?
[#161348809](https://www.pivotaltracker.com/story/show/161348809)

### CODE REVIEW

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -54,6 +54,14 @@ def get_queryset(self):
                 tag_list__name__icontains=self.request.GET['tag'])
         return articles
 
+    def article_read_time(self, article_body):
```
Comment (emugaya) [inline authors/apps/articles/views.py]: Rename this to ` read_time` and put it below create or as the last method in this class

**Review by emugaya — CHANGES_REQUESTED** [review]
Please add the expected `json` on the PR

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -69,6 +77,8 @@ def create(self, request, *args, **kwargs):
         if Article.objects.filter(slug=slug).exists():
             slug += str(time.time()).replace('.', '')
 
+        read = self.article_read_time(body)
```
Comment (emugaya) [inline authors/apps/articles/views.py]: Give this variable a better name.

File: authors/apps/articles/views.py [hunk 3]
```diff
@@ -140,7 +152,8 @@ def update(self, request, *args, **kwargs):
                 [str(tag) for tag in article.tag_list.all()]
             ),
             "image_url": request.data.get('image_url', article.image_url),
-            "audio_url": request.data.get('audio_url', article.audio_url)
+            "audio_url": request.data.get('audio_url', article.audio_url),
+            "read_time": request.data.get('read_time', article.read_time)
```
Comment (emugaya) [inline authors/apps/articles/views.py]: What happens when a user updates an article. Do you recalculate the read time?

File: authors/apps/articles/tests/Basetest.py [hunk 4]
```diff
@@ -36,7 +36,8 @@ def setUp(self):
             "tag_list": ["dragons", "fantacy"]
         }
         self.article_with_missing_fields = {
-            "title": "Getting started with JS"
+            "title": "Getting started with JS",
```
Comment (emugaya) [inline authors/apps/articles/tests/Basetest.py:41]: Add tests for read_time use a random string generator or faker library to generate body

**Review by brucemakallan — CHANGES_REQUESTED** [review]
Please add an example Response (in JSON) so we can see how read-time is displayed.

File: authors/apps/articles/views.py (hunk 1 above)
Comment (dnuwa) [reply authors/apps/articles/views.py]: This has been worked on

File: authors/apps/articles/views.py (hunk 2 above)
Comment (dnuwa) [reply authors/apps/articles/views.py]: The variable name has been changed readtime

File: authors/apps/articles/views.py (hunk 3 above)
Comment (dnuwa) [reply authors/apps/articles/views.py]: Thanx for pointing this out. The read time has to be recalculated upon saving the article changes. updates on the same have been made.

**Comment by emugaya** [comment]
@dnuwa I see you're returning a string as the read time. What if it's more than 60 minutes?

File: authors/apps/articles/tests/Basetest.py (hunk 4 above)
Comment (dnuwa) [reply authors/apps/articles/tests/Basetest.py]: > @dnuwa I see you're returning a string as the read time. What if it's more than 60 minutes?

The read time to all articles is returned in minutes even when it's beyond 60.

**Comment by dnuwa** [comment]
> @dnuwa I see you're returning a string as the read time. What if it's more than 60 minutes?

The read time has been changed to an integer value such that the API user can be able to change the minutes' value returned into hours.

File: authors/apps/articles/tests/Basetest.py (hunk 4 above)
Comment (dnuwa) [reply authors/apps/articles/tests/Basetest.py:41]: > Add tests for read_time use a random string generator or faker library to generate body

A faker library has been used to cater for this.

**Review by RachaelNantale — APPROVED** [review]
LGTM

**Review by brucemakallan — APPROVED** [review]

**Review by Sulaiman-Mozes — APPROVED** [review]

File: authors/apps/articles/migrations/0005_article_read_time.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        ('articles', '0004_merge_20181113_1623'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='article',
+            name='read_time',
+            field=models.IntegerField(default=0, max_length=100),
+        ),
+    ]
```
Comment (judeinno) [inline authors/apps/articles/migrations/0005_article_read_time.py:18]: Nice work on working on migrations feedback

File: authors/apps/articles/tests/Basetest.py [hunk 6]
```diff
@@ -22,7 +24,7 @@ def setUp(self):
             "title": "How to train your Dragon",
             # "slug": "How to train your Dragon",
             "description": "Ever wonder how?",
-            "body": "It takes a Jacobian",
+            "body": fake.text() * 10,
```
Comment (judeinno) [inline authors/apps/articles/tests/Basetest.py:27]: I liken that your using the faker library for the dummy data

**Review by judeinno — APPROVED** [review]
I approve

**Review by emugaya — APPROVED** [review]

**Pull request merged**